### PR TITLE
User profile

### PR DESCRIPTION
**Pull request by andru47**

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 1]
```diff
@@ -13,19 +14,31 @@
...
     String origin = request.getParameter("origin");
+    
+    if (userService.isUserLoggedIn() && userStorageObject.getUser(userService.getCurrentUser().getUserId()) == null) {
+      userStorageObject.addOrUpdateUser(new User(userService.getCurrentUser().getUserId(),
+                                           userService.getCurrentUser().getEmail(),
+                                           null,
+                                           new ArrayList<>(),
+                                           new ArrayList<>(),
+                                           new ArrayList<>()));
+    }
```
Comment (marcbury-g) [inline src/main/java/com/google/sps/servlets/AuthServlet.java]: Optional: Consider adding a builder for the User class similar as for the Event class.

File: src/main/java/com/google/sps/servlets/UserServlet.java [hunk 2]
```diff
@@ -16,43 +16,107 @@
...
 
   @Override
-  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    UserService userService = UserServiceFactory.getUserService();
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
     if (!userService.isUserLoggedIn()) {
       response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
       return;
     }
+    if (request.getPathInfo().equals("/username")) {
```
Comment (marcbury-g) [inline src/main/java/com/google/sps/servlets/UserServlet.java:91]: How do you get this information? Maybe have two paths in the doGet method: /users/events and /users/username ?

File: src/main/java/com/google/sps/servlets/UserServlet.java [hunk 3]
```diff
@@ -16,43 +16,107 @@
...
+
+  private List<Event> getEventsFromIds(List<String> idList) {
+    List<Event> eventList = new ArrayList<>();
+    for (String id : idList) {
+      Event queriedEvent = eventStorageObject.getEvent(id);
+      if (queriedEvent == null)
+        continue;
+      eventList.add(queriedEvent);
+    }
+    return eventList;
```
Comment (marcbury-g) [inline src/main/java/com/google/sps/servlets/UserServlet.java]: Add a TODO to batch the requests.

File: src/main/webapp/script.js [hunk 4]
```diff
@@ -347,3 +353,86 @@ function addPerson() {
 function setMinDateToToday() {
   document.getElementById("event-start-date").min = new Date().toISOString().slice(0, 10);
 }
+
+function fetchUserInfo() {
+  fetch("/users").then(handleError).then(response => response.json()).then(userInfo => {
+    console.log(userInfo);
+    const email = userInfo.email;
+    var username = userInfo.username;
```
Comment (marcbury-g) [inline src/main/webapp/script.js:360]: I do not see where you return these values in the backend code.

File: src/test/java/com/google/sps/servlets/TestAuthServlet.java [hunk 5]
```diff
@@ -55,9 +63,24 @@ public void testAuthServlet_doGet_WithLoggedOutUser_ReturnsLoginLink() throws IO
...
+    when(mockRequest.getParameter("origin")).thenReturn("someOrigin");
+    when(mockUserService.getCurrentUser()).thenReturn(new User("email", "domain", "user1"));
+    when(mockUserStorage.getUser("user1")).thenReturn(null);
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(mockResponse.getWriter()).thenReturn(writer);
+
+    new AuthServlet(mockUserService, mockUserStorage).doGet(mockRequest, mockResponse);
+
+    verify(mockUserStorage, atLeast(1)).addOrUpdateUser(any());
```
Comment (marcbury-g) [inline src/test/java/com/google/sps/servlets/TestAuthServlet.java]: Also check that the parameter is correct:
https://www.javadoc.io/doc/org.mockito/mockito-core/2.6.9/org/mockito/ArgumentCaptor.html

File: src/test/java/com/google/sps/servlets/TestUserServlet.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+import com.google.sps.data.*;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+public class TestUserServlet {
+  com.google.sps.data.User fakeUser;
```
Comment (marcbury-g) [inline src/test/java/com/google/sps/servlets/TestUserServlet.java]: Make this variable local in the methods.

File: src/main/java/com/google/sps/servlets/AuthServlet.java (hunk 1 above)
Comment (LadyPython) [inline src/main/java/com/google/sps/servlets/AuthServlet.java]: Maybe add builder and use .withId(userService.getCurrentUser().getUserId()).withEmail(userService.getCurrentUser().getEmail()).build()

File: src/main/webapp/user.html [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+          <div id="username-placeholder">
+          </div>
+          <div id="username-setter" style="display: none;">
+            <form action="/users/username" method="POST">
+              <input type="text" placeholder="Enter a desired nickname" name="nickname" required>
+              <input type="submit" />
+            </form>
+          </div>
+        </div>
+        <div id="joined-events"><button class="btn btn-success" onclick="changePopoverColorTo('#28a745')">Click here to see all the events you joined</button>
```
Comment (LadyPython) [inline src/main/webapp/user.html:51]: Maybe separate it to CSS?

**Review by LadyPython — APPROVED** [review]

File: src/test/java/com/google/sps/servlets/TestUserServlet.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        Arrays.asList(new String[] { "secondId" }), null);
+    when(mockUserService.isUserLoggedIn()).thenReturn(true);
+    when(mockRequest.getParameter("nickname")).thenReturn("andru47");
+    when(mockUserService.getCurrentUser()).thenReturn(new User("email", "authDomain", "123"));
+    when(mockUserStorage.getUser("123")).thenReturn(fakeUser);
+    when(mockRequest.getPathInfo()).thenReturn("/username");
+
+    new UserServlet(mockUserStorage, mockEventStorage, mockUserService).doPost(mockRequest, mockResponse);
+
+    assertTrue(fakeUser.getUsername().equals("andru47"));
```
Comment (marcbury-g) [inline src/test/java/com/google/sps/servlets/TestUserServlet.java:128]: You should check whether `userStorageObject.addOrUpdateUser` with the right argument is called.

File: src/main/webapp/user.html (hunk 7 above)
Comment (andru47) [reply src/main/webapp/user.html:51]: I can't add that to css. Each button has its own color and this color is for the tooltips. I haven't found other nicer way to make different colors to different tooltips

File: src/main/java/com/google/sps/servlets/UserServlet.java (hunk 2 above)
Comment (andru47) [reply src/main/java/com/google/sps/servlets/UserServlet.java:91]: Do we actually need another path for the post request? I have just realised that this servlet has only one job for the get request and one for the post

File: src/main/java/com/google/sps/servlets/UserServlet.java (hunk 3 above)
Comment (andru47) [reply src/main/java/com/google/sps/servlets/UserServlet.java]: I haven't considered to make the batching here, but will add a TODO

File: src/main/webapp/script.js (hunk 4 above)
Comment (andru47) [reply src/main/webapp/script.js:360]: https://github.com/googleinterns/step244-2020/blob/307919adfff0c6f32a9d59231c584f05f31170b3/src/main/java/com/google/sps/servlets/UserServlet.java#L72 I create a json object for user and then append to it the events from each type

File: src/main/java/com/google/sps/servlets/UserServlet.java (hunk 2 above)
Comment (marcbury-g) [reply src/main/java/com/google/sps/servlets/UserServlet.java:91]: If you don't need it then it is fine :)

**Review by marcbury-g — APPROVED** [review]

**Pull request merged**